### PR TITLE
Support connection break detection for e/poll

### PR DESCRIPTION
**Pull request by 13hakta**
Detect if connection was closed by remote side. Support poll and epoll pollers.

### CODE REVIEW

**Comment by resetius** [comment]
Could you pls add a test to https://github.com/resetius/coroio/blob/master/tests/tests.cpp for epoll/poll only
Thank you.

**Comment by 13hakta** [comment]
Added tests upon poll / epoll.

File: coroio/base.hpp [hunk 1]
```diff
@@ -27,13 +27,15 @@ struct TTimer {
...
+    THandle RHup;
 };
 
 struct TEvent {
     int Fd;
     enum {
         READ = 1,
-        WRITE = 2
+        WRITE = 2,
+        RHUP = 3
```
Comment (resetius) [inline coroio/base.hpp:38]: 4

File: tests/tests.cpp [hunk 2]
```diff
@@ -1056,6 +1056,41 @@ void test_uring_no_sqe(void** ) {
     }
 }
 
+template<typename TPoller>
+void test_remote_disconnect(void**) {
+    bool _changed = false;
```
Comment (resetius) [inline tests/tests.cpp:1061]: naming: pls don't use underscores.
```
bool changed
```
better

File: tests/tests.cpp [hunk 3]
```diff
@@ -1056,6 +1056,41 @@ void test_uring_no_sqe(void** ) {
...
+    TLoop loop;
+    TSocket socket(TAddress{"127.0.0.1", 8888}, loop.Poller());
+    TSocket clientSocket_src{};
+    TSocket clientSocket_dst{};
+    socket.Bind();
+    socket.Listen();
+
+    TFuture<void> h1 = [&clientSocket_src](TPoller& poller, bool *changed) -> TFuture<void>
+    {
+        clientSocket_src = TSocket(TAddress{"127.0.0.1", 8888}, poller);
```
Comment (resetius) [inline tests/tests.cpp:1073]: pls use simple local variable
```
auto clientSocket = ....
```

File: tests/tests.cpp [hunk 4]
```diff
@@ -1056,6 +1056,41 @@ void test_uring_no_sqe(void** ) {
...
+        co_await clientSocket_src.Connect();
+        co_await clientSocket_src.Monitor();
+        *changed = true;
+        co_return;
+    }(loop.Poller(), &_changed);
+
+    TFuture<void> h2 = [](TSocket* socket, TSocket* clientSocket) -> TFuture<void>
+    {
+        *clientSocket = std::move(co_await socket->Accept());
```
Comment (resetius) [inline tests/tests.cpp:1082]: pls use simple local variable
```
auto clientSocket = ... Accept()
```

**Comment by resetius** [comment]
What happens if I start a coroutine that waits on Monitor, and then, in a separate coroutine, I close the socket? Will the coroutine waiting on Monitor ever wake up, or could it remain suspended indefinitely after the socket is closed?

**Comment by 13hakta** [comment]
> What happens if I start a coroutine that waits on Monitor, and then, in a separate coroutine, I close the socket? Will the coroutine waiting on Monitor ever wake up, or could it remain suspended indefinitely after the socket is closed?

Closing connection on opposite side will cause RDHUP event and trigger Monitor() to finish, but closing on own side will trigger HUP event and remain suspended indefinitely.
BTW Monitor can be extended to support both events and return value of trigger event.

**Comment by resetius** [comment]
Could you pls remove "win32 support" patches. This work will be applied in a separate PR.
Als pls fix compilation issues. Thank you!

**Comment by resetius** [comment]
Pls fix assert in uring.cpp:107
```
- assert(ev.Type == (TEvent::READ|TEvent::WRITE));
+ assert(ev.Type == (TEvent::READ|TEvent::WRITE|TEvent::RHup));
```

**Pull request merged**